sdl_viewer/src/terrain_drawer/mod.rs

```diff
@@ -1,5 +1,193 @@
+        // from (0, 0, 0), the first call to camera_changed() will be very resource intensive
+        let u_transform = GlUniform::new(&program, "world_to_gl", Matrix4::identity());
+
+        let camera_pos_xy_m = Vector2::zero();
+
+        let vertex_array = GlVertexArray::new(Rc::clone(&gl));
+
+        let (buffer_position, buffer_indices, num_indices) =
+            Self::create_mesh(&program, &vertex_array, Rc::clone(&gl));
+
+        let terrain_layers = terrain_paths
+            .map(|p| TerrainLayer::new(&program, p, GRID_SIZE + 1).unwrap())
+            .collect();
+
+        Self {
+            program,
+            u_transform,
+            camera_pos_xy_m,
+            vertex_array,
+            buffer_position,
+            buffer_indices,
+            num_indices,
+            terrain_layers,
+        }
+    }
+
+    fn create_mesh(
+        program: &GlProgram,
+        vertex_array: &GlVertexArray,
+        gl: Rc<opengl::Gl>,
+    ) -> (GlBuffer, GlBuffer, usize) {
+        let num_vertices = (GRID_SIZE + 1) as usize * (GRID_SIZE + 1) as usize * 3;
+        let mut vertices: Vec<i32> = Vec::with_capacity(num_vertices);
+        for iy in 0..=GRID_SIZE as i32 {
+            for ix in 0..=GRID_SIZE as i32 {
+                vertices.push(ix);
+                vertices.push(iy);
+                vertices.push(0);
+            }
+        }
+
+        let flat_ix = |x: GLuint, y: GLuint| y * (GRID_SIZE + 1) as GLuint + x;
+        let mut indices: Vec<GLuint> =
+            Vec::with_capacity(GRID_SIZE as usize * GRID_SIZE as usize * 3 * 2);
+        for iy in 0..GRID_SIZE as GLuint {
+            for ix in 0..GRID_SIZE as GLuint {
+                // Two triangles = 1 quad
+                indices.push(flat_ix(ix, iy));
+                indices.push(flat_ix(ix + 1, iy));
+                indices.push(flat_ix(ix, iy + 1));
+                indices.push(flat_ix(ix + 1, iy));
+                indices.push(flat_ix(ix, iy + 1));
+                indices.push(flat_ix(ix + 1, iy + 1));
+            }
+        }
+
+        vertex_array.bind();
+
+        let buffer_position = GlBuffer::new_array_buffer(Rc::clone(&gl));
+        let buffer_indices = GlBuffer::new_element_array_buffer(Rc::clone(&gl));
+
+        buffer_position.bind();
+        unsafe {
+            program.gl.BufferData(
+                opengl::ARRAY_BUFFER,
+                (vertices.len() * mem::size_of::<i32>()) as GLsizeiptr,
+                vertices.as_ptr() as *const c_void,
+                opengl::STATIC_DRAW,
+            );
+
+            let pos_attr = gl.GetAttribLocation(program.id, c_str!("aPos"));
+            gl.EnableVertexAttribArray(pos_attr as GLuint);
+            // aPos is an ivec3. If we wanted, we could make it an ivec2 or a single index.
+            gl.VertexAttribIPointer(
+                pos_attr as GLuint,
+                3,
+                opengl::INT,
```

Comment:
noob question: is there a difference between `opengl::types::GLint` and `opengl::INT` or `opengl::types::GLuint` and `opengl::UNSIGNED_INT`?

Reply:
Yes, `GLint` is a type and `opengl::INT` is an enum/constant value that represents the type